python/ray/experimental/data/dataset.py

```diff
@@ -869,18 +869,65 @@ def to_torch(self, **todo) -> "torch.utils.data.IterableDataset":
         """
         raise NotImplementedError  # P1
 
-    def to_tf(self, **todo) -> "tf.data.Dataset":
-        """Return a TF data iterator over this dataset.
+    def to_tf(self,
+              label_column: str,
+              output_signature: List["tf.TypeSpec"],
+              feature_columns: Optional[List[str]] = None,
+              prefetch_blocks: int = 0,
+              batch_size: int = None) -> "tf.data.Dataset":
+        """Return a TF Dataset over this dataset.
+
+        The TF Dataset will be created from the generator returned by the
+        ``iter_batches`` method. ``prefetch_blocks`` and ``batch_size``
+        arguments will be passed to that method.
+
+        This is only supported for datasets convertible to Arrow records.
+
+        Requires all datasets to have the same columns.
 
         Note that you probably want to call ``.split()`` on this dataset if
         there are to be multiple TensorFlow workers consuming the data.
 
+        The elements generated must be compatible with the given
+        ``output_signature`` argument (same as in
+        ``tf.data.Dataset.from_generator``).
+
         Time complexity: O(1)
 
+        Args:
+            label_column (str): The name of the column used as the label
+                (second element of the output tuple).
+            output_signature (List[tf.TypeSpec]): A 2-element list
+                of `tf.TypeSpec` objects corresponding to (features, label).
+            feature_columns (Optional[List[str]]): List of columns in datasets
+                to use. If None, all columns will be used.
+            prefetch_blocks: The number of blocks to prefetch ahead of the
+                current block during the scan.
+            batch_size: Record batch size, or None to let the system pick.
+
         Returns:
             A tf.data.Dataset.
         """
-        raise NotImplementedError  # P1
+
+        # argument exception checking is done in from_generator
+
+        try:
+            import tensorflow as tf
+        except ImportError:
+            raise ValueError("tensorflow must be installed!")
+
+        def make_generator():
+            for batch in self.iter_batches(
+                    prefetch_blocks=prefetch_blocks,
+                    batch_size=batch_size,
+                    batch_format="pandas"):
```

Comment:
I don't think it makes sense to expose the format here as it will need to be converted to numpy arrays to work with tensorflow anyway

Reply:
I was thinking that converting directly from Arrow Tables to TensorFlow Tensors might be nicer than the Arrow Table --> Pandas DataFrame --> NumPy ndarrays --> TensorFlow Tensors conversion, but this should all be zero-copy anyway so it shouldn't matter much. Just to give an idea, this is what it would look like when operating directly on Arrow Tables:
```python
for batch in self.iter_batches(
        prefetch_blocks=prefetch_blocks,
        batch_size=batch_size,
        batch_format="pyarrow"):
    target_col = batch.column(label_column)
    if feature_columns:
        batch = batch.select(feature_columns)
    yield batch.columns, target
```
I _think_, since `pyarrow.ChunkedArray` implements the `__array__` protocol, that an explicit conversion to NumPy wouldn't be necessary and would instead happen [implicitly within TensorFlow](https://github.com/tensorflow/tensorflow/blob/a4dfb8d1a71385bd6d122e4f27f86dcebb96712d/tensorflow/python/framework/tensor_util.py#L435).

Reply:
Gotcha, makes sense. I can try it with Arrow, it would indeed be nicer

Reply:
Tried it. It changes the shape of the resulting tensors. I think it'd be best to just go with pandas for now, and we can change it for the beta if needed

Reply:
Hmm interesting! 🤔  Going with Pandas sounds good. Thanks for trying it out!